File: apps/api/src/oneepis_api/models/clinical_record.py
from __future__ import annotations

import uuid
from datetime import date, datetime
from typing import TYPE_CHECKING, Any

from sqlalchemy import JSON, Date, DateTime, Enum, ForeignKey, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from oneepis_api.db.base import Base
from oneepis_api.models.appointment import AppointmentStatus as AppointmentStatus
from oneepis_api.models.appointment import ClinicalAppointment as ClinicalAppointment
from oneepis_api.models.base import IdMixin, TimestampMixin
from oneepis_api.models.clinical_enums import (
    AllergySeverity,
    ClinicalEntryKind,
    ClinicalEntryStatus,
    ClinicalEventSourceType,
    ClinicalEventType,
    EncounterStatus,
    EncounterType,
    EncounterWorkflowKind,
    RecordStatus,
)
from oneepis_api.models.medication_catalog import MedicationCatalogItem
from oneepis_api.models.patient import enum_values
from oneepis_api.models.vital_sign import VitalSign as VitalSign
from oneepis_api.services.medication_read_support import (
    medication_missing_fields,
    medication_source,
)

if TYPE_CHECKING:
    from oneepis_api.models.patient import Patient


class ClinicalEntry(Base, IdMixin, TimestampMixin):
    __tablename__ = "clinical_entries"

    patient_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("patients.id", ondelete="CASCADE"),
        index=True,
    )
    encounter_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("clinical_encounters.id", ondelete="SET NULL"),
        index=True,
        nullable=True,
    )
    kind: Mapped[ClinicalEntryKind] = mapped_column(
        Enum(ClinicalEntryKind, values_callable=enum_values, name="clinical_entry_kind"),
        nullable=False,
    )
    status: Mapped[ClinicalEntryStatus] = mapped_column(
        Enum(ClinicalEntryStatus, values_callable=enum_values, name="clinical_entry_status"),
        default=ClinicalEntryStatus.DRAFT,
        nullable=False,
    )
    occurred_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), index=True, nullable=False
    )
    title: Mapped[str] = mapped_column(String(160), nullable=False)
    subjective: Mapped[str | None] = mapped_column(Text, nullable=True)
    objective: Mapped[str | None] = mapped_column(Text, nullable=True)
    assessment: Mapped[str | None] = mapped_column(Text, nullable=True)
    plan: Mapped[str | None] = mapped_column(Text, nullable=True)
    tags: Mapped[list[str]] = mapped_column(JSON, default=list, nullable=False)
    extra_data: Mapped[dict[str, Any]] = mapped_column(
        "metadata", JSON, default=dict, nullable=False
    )
    created_by: Mapped[str] = mapped_column(String(120), nullable=False, default="system")

    patient: Mapped[Patient] = relationship(back_populates="clinical_entries")
    encounter: Mapped[ClinicalEncounter | None] = relationship(
        back_populates="clinical_entries",
    )


class ClinicalEvent(Base, IdMixin, TimestampMixin):
    __tablename__ = "clinical_events"

    patient_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("patients.id", ondelete="CASCADE"),
        index=True,
    )
    encounter_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("clinical_encounters.id", ondelete="SET NULL"),
        index=True,
        nullable=True,
    )
    event_type: Mapped[ClinicalEventType] = mapped_column(
        Enum(ClinicalEventType, values_callable=enum_values, name="clinical_event_type"),
        nullable=False,
    )
    occurred_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), index=True, nullable=False
    )
    summary: Mapped[str] = mapped_column(String(280), nullable=False)
    source_type: Mapped[ClinicalEventSourceType] = mapped_column(
        Enum(
            ClinicalEventSourceType,
            values_callable=enum_values,
            name="clinical_event_source_type",
        ),
        default=ClinicalEventSourceType.MANUAL,
        nullable=False,
    )
    source_ref: Mapped[str | None] = mapped_column(String(160), nullable=True)
    payload: Mapped[dict[str, Any]] = mapped_column(JSON, default=dict, nullable=False)
    created_by: Mapped[str] = mapped_column(String(120), nullable=False, default="system")

    patient: Mapped[Patient] = relationship(back_populates="clinical_events")
    encounter: Mapped[ClinicalEncounter | None] = relationship(back_populates="clinical_events")


class Allergy(Base, IdMixin, TimestampMixin):
    __tablename__ = "allergies"

    patient_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("patients.id", ondelete="CASCADE"),
        index=True,
    )
    substance: Mapped[str] = mapped_column(String(160), nullable=False)
    reaction: Mapped[str | None] = mapped_column(String(240), nullable=True)
    severity: Mapped[AllergySeverity] = mapped_column(
        Enum(AllergySeverity, values_callable=enum_values, name="allergy_severity"),
        default=AllergySeverity.UNKNOWN,
        nullable=False,
    )
    status: Mapped[RecordStatus] = mapped_column(
        Enum(RecordStatus, values_callable=enum_values, name="record_status"),
        default=RecordStatus.ACTIVE,
        nullable=False,
    )
    recorded_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)

    patient: Mapped[Patient] = relationship(back_populates="allergies")


class Medication(Base, IdMixin, TimestampMixin):
    __tablename__ = "medications"

    patient_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("patients.id", ondelete="CASCADE"),
        index=True,
    )
    catalog_item_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("medication_catalog_items.id", ondelete="SET NULL"),
        index=True,
        nullable=True,
    )
    name: Mapped[str] = mapped_column(String(160), nullable=False)
    dose: Mapped[str | None] = mapped_column(String(120), nullable=True)
    route: Mapped[str | None] = mapped_column(String(80), nullable=True)
    frequency: Mapped[str | None] = mapped_column(String(120), nullable=True)
    status: Mapped[RecordStatus] = mapped_column(
        Enum(RecordStatus, values_callable=enum_values, name="record_status"),
        default=RecordStatus.ACTIVE,
        nullable=False,
    )
    started_on: Mapped[date | None] = mapped_column(Date, nullable=True)
    ended_on: Mapped[date | None] = mapped_column(Date, nullable=True)
    dose_check_snapshot: Mapped[dict[str, Any]] = mapped_column(
        JSON,
        default=dict,
        nullable=False,
    )
    dose_override_reason: Mapped[str | None] = mapped_column(String(280), nullable=True)

    patient: Mapped[Patient] = relationship(back_populates="medications")
    catalog_item: Mapped[MedicationCatalogItem | None] = relationship()

    @property
    def source(self) -> dict[str, Any] | None:
        return medication_source(self)

    @property
    def missing_fields(self) -> list[str]:
        return medication_missing_fields(self)


class ActiveProblem(Base, IdMixin, TimestampMixin):
    __tablename__ = "active_problems"

    patient_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("patients.id", ondelete="CASCADE"),
        index=True,
    )
    title: Mapped[str] = mapped_column(String(160), nullable=False)
    code_system: Mapped[str | None] = mapped_column(String(80), nullable=True)
    code: Mapped[str | None] = mapped_column(String(80), nullable=True)
    status: Mapped[RecordStatus] = mapped_column(
        Enum(RecordStatus, values_callable=enum_values, name="record_status"),
        default=RecordStatus.ACTIVE,
        nullable=False,
    )
    onset_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    resolved_on: Mapped[date | None] = mapped_column(Date, nullable=True)
    notes: Mapped[str | None] = mapped_column(String(280), nullable=True)

    patient: Mapped[Patient] = relationship(back_populates="active_problems")


class ClinicalEncounter(Base, IdMixin, TimestampMixin):
    __tablename__ = "clinical_encounters"

    patient_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("patients.id", ondelete="CASCADE"),
        index=True,
    )
    type: Mapped[EncounterType] = mapped_column(
        Enum(EncounterType, values_callable=enum_values, name="encounter_type"),
        default=EncounterType.UNKNOWN,
        nullable=False,
    )
    status: Mapped[EncounterStatus] = mapped_column(
        Enum(EncounterStatus, values_callable=enum_values, name="encounter_status"),
        default=EncounterStatus.IN_PROGRESS,
        nullable=False,
    )
    workflow_kind: Mapped[EncounterWorkflowKind] = mapped_column(
        Enum(
            EncounterWorkflowKind,
            values_callable=enum_values,
            name="encounter_workflow_kind",
        ),
        default=EncounterWorkflowKind.GENERAL,
        nullable=False,
    )
    reason: Mapped[str] = mapped_column(String(200), nullable=False)
    started_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        index=True,
        nullable=False,
    )
    ended_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    location_label: Mapped[str | None] = mapped_column(String(120), nullable=True)
    notes: Mapped[str | None] = mapped_column(String(320), nullable=True)

    patient: Mapped[Patient] = relationship(back_populates="encounters")
    clinical_entries: Mapped[list[ClinicalEntry]] = relationship(
        back_populates="encounter",
        passive_deletes=True,
    )
    clinical_events: Mapped[list[ClinicalEvent]] = relationship(
        back_populates="encounter",
        passive_deletes=True,
    )

File: apps/web/src/lib/type-contracts/clinical-record.ts
import type { Patient } from "./patient";

export type ClinicalEntryKind =
  | "intake"
  | "progress"
  | "discharge_summary"
  | "lab_result"
  | "prescription"
  | "procedure"
  | "note";
export type ClinicalEntryStatus = "draft" | "signed" | "amended";
export type ClinicalEventType =
  | "symptom"
  | "vital_sign"
  | "exam_result"
  | "diagnosis"
  | "medication"
  | "procedure"
  | "clinical_note"
  | "care_plan"
  | "administrative";
export type ClinicalEventSourceType = "manual" | "clinical_entry" | "vital_sign" | "imported_document" | "ai_draft";
export type RecordStatus = "active" | "inactive" | "resolved" | "entered_in_error";
export type AllergySeverity = "mild" | "moderate" | "severe" | "unknown";
export type EncounterType = "ambulatory" | "hospitalization" | "emergency" | "unknown";
export type EncounterStatus = "scheduled" | "in_progress" | "completed" | "cancelled";
export type MedicationReadSourceSystem =
  | "local_curated"
  | "fda_openfda_label"
  | "fda_drugsfda"
  | "fda_enforcement"
  | "fda_faers"
  | "isp_anamed";
export type MedicationReadSourceReviewStatus = "draft" | "reviewed" | "deprecated";
export type MedicationMissingField = "dose" | "route" | "frequency" | "source";
export type EncounterWorkflowKind =
  | "general"
  | "ambulatory_preconsult"
  | "ambulatory_visit"
  | "hospitalization_admission"
  | "hospitalization_daily"
  | "hospitalization_discharge";

export type ClinicalEntry = {
  id: string;
  patient_id: string;
  encounter_id?: string | null;
  kind: ClinicalEntryKind;
  status: ClinicalEntryStatus;
  occurred_at: string;
  title: string;
  subjective?: string | null;
  objective?: string | null;
  assessment?: string | null;
  plan?: string | null;
  tags: string[];
  created_by: string;
  created_at: string;
  updated_at: string;
};

export type ClinicalEntryCreate = {
  encounter_id?: string | null;
  kind: ClinicalEntryKind;
  status?: ClinicalEntryStatus;
  occurred_at: string;
  title: string;
  subjective?: string | null;
  objective?: string | null;
  assessment?: string | null;
  plan?: string | null;
  tags?: string[];
  extra_data?: Record<string, unknown>;
  created_by?: string;
};

export type ClinicalEntryUpdate = Partial<
  Pick<
    ClinicalEntryCreate,
    "status" | "occurred_at" | "title" | "subjective" | "objective" | "assessment" | "plan" | "tags"
  >
>;

export type ClinicalEvent = {
  id: string;
  patient_id: string;
  encounter_id?: string | null;
  event_type: ClinicalEventType;
  occurred_at: string;
  summary: string;
  source_type: ClinicalEventSourceType;
  source_ref?: string | null;
  payload: Record<string, unknown>;
  created_by: string;
  created_at: string;
  updated_at: string;
};

export type ClinicalEventCreate = {
  encounter_id?: string | null;
  event_type: ClinicalEventType;
  occurred_at: string;
  summary: string;
  source_type?: ClinicalEventSourceType;
  source_ref?: string | null;
  payload?: Record<string, unknown>;
  created_by?: string;
};

export type ClinicalEventUpdate = Partial<ClinicalEventCreate>;

export type ClinicalTimeline = {
  events: ClinicalEvent[];
  entries: ClinicalEntry[];
};
export type VitalSignCreate = {
  measured_at: string;
  temperature_c?: string | null;
  systolic_bp?: number | null;
  diastolic_bp?: number | null;
  heart_rate_bpm?: number | null;
  respiratory_rate_bpm?: number | null;
  oxygen_saturation_pct?: string | null;
  notes?: string | null;
};

export type VitalSign = {
  id: string;
  patient_id: string;
  measured_at: string;
  temperature_c?: string | null;
  systolic_bp?: number | null;
  diastolic_bp?: number | null;
  heart_rate_bpm?: number | null;
  respiratory_rate_bpm?: number | null;
  oxygen_saturation_pct?: string | null;
  notes?: string | null;
  created_at: string;
  updated_at: string;
};

export type VitalSignUpdate = Partial<VitalSignCreate>;

export type AllergyCreate = {
  substance: string;
  reaction?: string | null;
  severity?: AllergySeverity;
  status?: RecordStatus;
  recorded_at: string;
};

export type Allergy = {
  id: string;
  patient_id: string;
  substance: string;
  reaction?: string | null;
  severity: AllergySeverity;
  status: RecordStatus;
  recorded_at: string;
  created_at: string;
  updated_at: string;
};

export type AllergyUpdate = Partial<AllergyCreate>;

export type MedicationCreate = {
  catalog_item_id?: string | null;
  name: string;
  dose?: string | null;
  route?: string | null;
  frequency?: string | null;
  status?: RecordStatus;
  started_on?: string | null;
  ended_on?: string | null;
  dose_override_reason?: string | null;
};

export type MedicationSourceRead = {
  source_system: MedicationReadSourceSystem;
  source_label: string;
  source_url?: string | null;
  external_id?: string | null;
  source_version?: string | null;
  retrieved_at?: string | null;
  reviewed_at?: string | null;
  review_status: MedicationReadSourceReviewStatus;
};

export type MedicationRead = {
  id: string;
  patient_id: string;
  catalog_item_id?: string | null;
  name: string;
  dose?: string | null;
  route?: string | null;
  frequency?: string | null;
  status: RecordStatus;
  started_on?: string | null;
  ended_on?: string | null;
  dose_check_snapshot: Record<string, unknown>;
  dose_override_reason?: string | null;
  source?: MedicationSourceRead | null;
  missing_fields: MedicationMissingField[];
  created_at: string;
  updated_at: string;
};

export type Medication = MedicationRead;

export type MedicationUpdate = Partial<MedicationCreate>;

export type ActiveProblemCreate = {
  title: string;
  code_system?: string | null;
  code?: string | null;
  status?: RecordStatus;
  onset_date?: string | null;
  resolved_on?: string | null;
  notes?: string | null;
};

export type ActiveProblem = {
  id: string;
  patient_id: string;
  title: string;
  code_system?: string | null;
  code?: string | null;
  status: RecordStatus;
  onset_date?: string | null;
  resolved_on?: string | null;
  notes?: string | null;
  created_at: string;
  updated_at: string;
};

export type ActiveProblemUpdate = Partial<ActiveProblemCreate>;

export type HistoricalDiagnosis = {
  source_event_id: string;
  title: string;
  occurred_at: string;
  source_type: ClinicalEventSourceType;
  source_ref?: string | null;
  source_label: string;
  limit: string;
  code_system?: string | null;
  code?: string | null;
};

export type ClinicalEncounterCreate = {
  type?: EncounterType;
  status?: EncounterStatus;
  workflow_kind?: EncounterWorkflowKind;
  reason: string;
  started_at: string;
  ended_at?: string | null;
  location_label?: string | null;
  notes?: string | null;
};

export type ClinicalEncounter = {
  id: string;
  patient_id: string;
  type: EncounterType;
  status: EncounterStatus;
  workflow_kind: EncounterWorkflowKind;
  reason: string;
  started_at: string;
  ended_at?: string | null;
  location_label?: string | null;
  notes?: string | null;
  created_at: string;
  updated_at: string;
};

export type ClinicalEncounterUpdate = Partial<ClinicalEncounterCreate>;
export type PatientRecordSnapshot = {
  patient: Patient;
  latest_vitals?: VitalSign | null;
  active_allergies: Allergy[];
  active_medications: Medication[];
  active_problems: ActiveProblem[];
  historical_diagnoses: HistoricalDiagnosis[];
  recent_entries: ClinicalEntry[];
};

export type AuditEvent = {
  id: string;
  action: string;
  entity_type: string;
  entity_id?: string | null;
  actor_id: string;
  correlation_id?: string | null;
  request_method?: string | null;
  request_path?: string | null;
  extra_data: Record<string, unknown>;
  created_at: string;
};
